### PR TITLE
Port generic alpaka phi functions to `DataFormats/Math`

### PR DESCRIPTION
**Pull request by VourMa**
During the review of the LST integration PR #45117 it was identified that some of the math functions rewritten in the alpaka framework for usage in the LST algorithm could be made more widely available in CMSSW by porting them to a more central package:
- https://github.com/cms-sw/cmssw/pull/45117#discussion_r1674782508
- https://github.com/cms-sw/cmssw/pull/45117#discussion_r1749143439

This PR follows up on those comments. After the initial PR review, the choice made was to port these functions under `Heterogeneous/AlpakaMath`. The function logic is as close as possible to the non-alpaka equivalent `DataFormats/Math/interface/deltaPhi.h` file.

This PR only introduces the new function to CMSSW + the relevant unit tests. A follow-up PR will be made to actually use these functions in the LST algorithm and delete the LST-specific implementations.

### CODE REVIEW

**Comment by makortel** [comment]
@cms-sw/heterogeneous-l2 (not sure if this PR would need to be assigned to `heterogeneous` as well)

> The choice made was to port these functions under `DataFormats/Math`, creating a new folder, `interface/alpaka` but the exact placement could be discussed. The function logic is as close as possible to the non-alpaka equivalent `DataFormats/Math/interface/deltaPhi.h` file.

Thinking out loud
* Considering package dependencies, making `DataFormats/Math` to depend on Alpaka would add an unnecessary dependence on Alpaka to nearly all packages that presently depend on `DataFormats/Math` (376 direct dependents)
  * A proper dependence would need to be added to `DataFormats/Math/BuildFile.xml`
  * From dependency standpoint e.g. `HeterogeneousCore/AlpakaInterface` would be a natural placement
* Per https://github.com/cms-sw/cmssw/blob/master/HeterogeneousCore/AlpakaCore/README.md#overall-guidelines the header does not need to be in `interface/alpaka`, because it does not depend on `ALPAKA_ACCELERATOR_NAMESPACE`. The wording there doesn't preclude it though, although we have generally preferred to place such headers directly to `interface` directory.

File: DataFormats/Math/interface/alpaka/deltaPhi.h [hunk 1]
```diff
@@ -0,0 +1,35 @@
+#ifndef DataFormats_Math_alpaka_deltaPhi_h
+#define DataFormats_Math_alpaka_deltaPhi_h
+
+#include "HeterogeneousCore/AlpakaInterface/interface/config.h"
```
Comment (makortel) [inline DataFormats/Math/interface/alpaka/deltaPhi.h:4]:
```suggestion
#include <alpaka/alpaka.hpp>
```
would seem to be sufficient.

File: DataFormats/Math/interface/alpaka/deltaPhi.h [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+namespace cms::alpakatools {
+
+  // reduce to [-pi,pi]
+  template <typename TAcc, typename T>
+  ALPAKA_FN_HOST_ACC ALPAKA_FN_INLINE constexpr T reduceRange(TAcc const& acc, T x) {
```
Comment (makortel) [inline DataFormats/Math/interface/alpaka/deltaPhi.h:10]: I was going to comment the `ALPAKA_FN_HOST_ACC ALPAKA_FN_INLINE` are not necessary because the function is `constexpr`, but after looking at Alpaka implementation of e.g. `abs()` and `round()` 
https://github.com/alpaka-group/alpaka/blob/a4142d3feb7686d803e1ec5f25d7b2278337f455/include/alpaka/math/Traits.hpp#L864
https://github.com/alpaka-group/alpaka/blob/a4142d3feb7686d803e1ec5f25d7b2278337f455/include/alpaka/math/Traits.hpp#L1331
that are **not** `constexpr`, I'm actually puzzled how this function can be `constexpr`.

File: DataFormats/Math/interface/alpaka/deltaPhi.h [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+    constexpr T o2pi = 1. / (2. * M_PI);
```
Comment (makortel) [inline DataFormats/Math/interface/alpaka/deltaPhi.h:11]: Now that we have C++20, how about
```suggestion
    constexpr T o2pi = T{1.} / (T{2.} * std::numbers::pi_v<T>);
```
?

Is it intentional to do the calculations on double and cast the result to `T`?

Is `T` expected to be only a floating point type?

**Review by makortel — COMMENTED** [review]
I'd suggest to add unit tests.

File: DataFormats/Math/interface/alpaka/deltaPhi.h (hunk 3 above)
Comment (slava77) [reply DataFormats/Math/interface/alpaka/deltaPhi.h:11]: I think that it would be better to maintain the implementation as close as possible to the original `deltaPhi.h`

**Comment by slava77** [comment]
> Considering package dependencies, making `DataFormats/Math` to depend on Alpaka would add an unnecessary dependence on Alpaka to nearly all packages that presently depend on `DataFormats/Math` (376 direct dependents)
> 
>     * A proper dependence would need to be added to `DataFormats/Math/BuildFile.xml`

is there a way to keep the dependence only via `/alpaka` directory includes? ... I guess the question goes in the direction of having it in a separate directory, effectively.

**Comment by makortel** [comment]
> > Considering package dependencies, making `DataFormats/Math` to depend on Alpaka would add an unnecessary dependence on Alpaka to nearly all packages that presently depend on `DataFormats/Math` (376 direct dependents)
> > ```
> > * A proper dependence would need to be added to `DataFormats/Math/BuildFile.xml`
> > ```
> 
> is there a way to keep the dependence only via `/alpaka` directory includes? ... I guess the question goes in the direction of having it in a separate directory, effectively.

No. For link-time dependencies that wouldn't work. Some source-level operations, like IIRC `git cms-checkdeps -a`, operate based on what files have been modified.

**Comment by VourMa** [comment]
> Thinking out loud
> 
>   * Considering package dependencies, making `DataFormats/Math` to depend on Alpaka would add an unnecessary dependence on Alpaka to nearly all packages that presently depend on `DataFormats/Math` (376 direct dependents)
>       
>     * A proper dependence would need to be added to `DataFormats/Math/BuildFile.xml`
>     * From dependency standpoint e.g. `HeterogeneousCore/AlpakaInterface` would be a natural placement. We could also think of a new package, e.g. `AlpakaMath`.
> 
>   * Per https://github.com/cms-sw/cmssw/blob/master/HeterogeneousCore/AlpakaCore/README.md#overall-guidelines the header does not need to be in `interface/alpaka`, because it does not depend on `ALPAKA_ACCELERATOR_NAMESPACE`. The wording there doesn't preclude it though, although we have generally preferred to place such headers directly to `interface` directory.

Based on this and the follow up discussion, I understand that the current placement is not ideal, since the dependencies cannot be reduced. During our internal review, I had this initially placed under a new `DataFormats` subpackage, so I revert and put it under `DataFormats/AlpakaMath/interface` (i.e. also dropping the `alpaka` directory).

**Comment by VourMa** [comment]
> I'd suggest to add unit tests.

Sure. Do you have in mind some unit test that I can use as a template (to speed up the work and to make sure that I cover what needs to be tested)?

**Comment by slava77** [comment]
> Based on this and the follow up discussion, I understand that the current placement is not ideal, since the dependencies cannot be reduced. During our internal review, I had this initially placed under a new `DataFormats` subpackage, so I revert and put it under `DataFormats/AlpakaMath/interface` (i.e. also dropping the `alpaka` directory).

IIUC, `HeterogeneousCore/` subsystem was mentioned earlier by Matti, not `DataFormats`.

**Comment by VourMa** [comment]
> > Based on this and the follow up discussion, I understand that the current placement is not ideal, since the dependencies cannot be reduced. During our internal review, I had this initially placed under a new `DataFormats` subpackage, so I revert and put it under `DataFormats/AlpakaMath/interface` (i.e. also dropping the `alpaka` directory).
> 
> IIUC, `HeterogeneousCore/` subsystem was mentioned earlier by Matti, not `DataFormats`.

Ah, sorry, I misinterpreted the comment to mean that the new subpackage would be under `DataFormats`. It's clear now, thanks.

**Comment by slava77** [comment]
> Ah, sorry, I misinterpreted the comment to mean that the new subpackage would be under `DataFormats`. It's clear now, thanks.

I guess @makortel could confirm the preference to avoid further confusion.

File: DataFormats/Math/interface/alpaka/deltaPhi.h (hunk 3 above)
Comment (makortel) [reply DataFormats/Math/interface/alpaka/deltaPhi.h:11]: > That said, your suggestion makes sense but maybe it should be propagated to `interface/deltaPhi.h` as well? Not sure whether we want to do so though (it was not the initial purpose of this PR).

Ok, maybe it would make sense to keep both versions consistent, and maybe any changes would be beyond this PR.

> I suppose that the intention is that `T` is only a floating point type but I am not sure. For LST purposes (for which these functions are going to be used, at least initially), floating point type only `T` is enough.

I was wondering if anyone would be interested to add that as a type constraint either as
```cpp
  template <typename TAcc, typename T>
  requires std::is_floating_point_v<T>
  ALPAKA_FN_HOST_ACC ALPAKA_FN_INLINE T reduceRange(TAcc const& acc, T x) {
```
or
```cpp
  template <typename TAcc>
  ALPAKA_FN_HOST_ACC ALPAKA_FN_INLINE std::floating_point auto reduceRange(TAcc const& acc, std::floating_point auto x) {
```
but probably that's beyond this PR as well.

**Comment by makortel** [comment]
> > I'd suggest to add unit tests.
> 
> Sure. Do you have in mind some unit test that I can use as a template (to speed up the work and to make sure that I cover what needs to be tested)?

By quick look I see the existing `reco::deltaPhi()` and `reco::reduceRange()` being used in assertive way in
https://github.com/cms-sw/cmssw/blob/f55e6a58353c43d861d68322d73661429df07810/DataFormats/Math/test/testAtan2.cpp#L187-L189
(and even there it is not immediately clear whether `reduceRange()` or `int2phi()` is the one being tested).

I'd say the primary objective would be to ensure all functions can be called for the relevant data types (I suppose both `float` and `double`) can be used in the kernel code for all backends, and ideally at least for some inputs the functions return expected values (trying to cover all corner cases is a potential rabbit hole though).

Here are a few rather simple Alpaka-using unit tests
* https://github.com/cms-sw/cmssw/blob/master/HeterogeneousCore/AlpakaInterface/test/alpaka/testAtomicPairCounter.dev.cc
* https://github.com/cms-sw/cmssw/blob/master/MagneticField/ParametrizedEngine/test/alpaka/testParabolicParametrizedMagneticField.dev.cc

I personally prefer the Catch2 unit test framework (used in `testAtomicPairCounter.dev.cc`), but that is not required.

**Comment by makortel** [comment]
> > Ah, sorry, I misinterpreted the comment to mean that the new subpackage would be under `DataFormats`. It's clear now, thanks.
> 
> I guess @makortel could confirm the preference to avoid further confusion.

I'd wait for next week before moving things around to get more feedback (after more people will presumably be back)

I would not be against `DataFormats/AlpakaMath`, but I'd ask the questions
* What is the probability these functions will be used in DataFormats code?
  * If guessed to be at most tiny, the package could well be placed in other subsystem to avoid placing non-DataFormat code in DataFormats
* What is the probability there will be more Alpaka-based math algorithms?
  * If guessed at most tiny, perhaps an existing package would be less of a burden than a new package for one header
  * If guessed to be above tiny (I could imagine this to be the case), a new package could be justified

File: DataFormats/Math/interface/alpaka/deltaPhi.h (hunk 2 above)
Comment (slava77) [reply DataFormats/Math/interface/alpaka/deltaPhi.h:10]: @makortel when do we expect to switch to C++23? These functions will be constexpr in the standard STL at that point.

Comment (VourMa) [reply DataFormats/Math/interface/alpaka/deltaPhi.h:10]: Indeed, you are right. Compiling and using the functions did not show any errors though, is it that `constexpr` is silently ignored? If that's the case, I guess I should remove it so that it does not cause confusion.

Comment (makortel) [reply DataFormats/Math/interface/alpaka/deltaPhi.h:10]: > when do we expect to switch to C++23?

Based on the experience with C++20, maybe around 2027.

Comment (makortel) [reply DataFormats/Math/interface/alpaka/deltaPhi.h:10]: > Indeed, you are right. Compiling and using the functions did not show any errors though, is it that `constexpr` is silently ignored? If that's the case, I guess I should remove it so that it does not cause confusion.

It seems that `constexpr` functions and function templates are allowed to call non-`constexpr` functions as long as there is at least one set of arguments with which the function can be evaluated as a constant expression (and otherwise the program is ill-formed, no diagnostic required...).

From C++20 draft ([N4849](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2020/n4849.pdf)), 9.2.5.7 (.6 is also relevant)
> If the instantiated template specialization of a constexpr function template or member function of a class template would fail to satisfy the requirements for a constexpr function, that specialization is still a constexpr function, even though a call to such a function cannot appear in a constant expression. If no specialization of the template would satisfy the requirements for a constexpr function when considered as a non-template function, the template is ill-formed, no diagnostic required.

So while our compilers might not warn about the situation, C++20-conforming code would not use `constexpr` here. Although IIUC this constraint was removed from C++23.

**Comment by VourMa** [comment]
> I would not be against `DataFormats/AlpakaMath`, but I'd ask the questions
> 
>   * What is the probability these functions will be used in DataFormats code?
>       
>     * If guessed to be at most tiny, the package could well be placed in other subsystem to avoid placing non-DataFormat code in DataFormats

My estimation is that it's tiny, so I will not go with that option.

>   * What is the probability there will be more Alpaka-based math algorithms?
>       
>     * If guessed at most tiny, perhaps an existing package would be less of a burden than a new package for one header
>     * If guessed to be above tiny (I could imagine this to be the case), a new package could be justified

I cannot really predict and I don't have a use case right now, but I can see people needing, e.g., `deltaR` in Alpaka code sooner than later. Based on that, I would go ahead and make a new `AlpakaMath` package. I will do so tomorrow, unless someone comes up with a strong preference.

File: DataFormats/Math/interface/alpaka/deltaPhi.h (hunk 2 above)
Comment (VourMa) [reply DataFormats/Math/interface/alpaka/deltaPhi.h:10]: Since C++23 seems to be quite a bit away for CMSSW, I would go with the C++20 conforming option and remove the `constexpr`.

**Comment by slava77** [comment]
@cmsbuild please test

I'm not sure why this was left unattended from the review side for two days

**Comment by fwyzard** [comment]
> I'm not sure why this was left unattended from the review side for two days

On my side, because I am busy with other tasks.

File: HeterogeneousCore/AlpakaMath/interface/deltaPhi.h [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+#define HeterogeneousCore_AlpakaMath_deltaPhi_h
+
+#include <alpaka/alpaka.hpp>
+
+namespace cms::alpakatools {
+
+  // reduce to [-pi,pi]
+  template <typename TAcc, typename T>
+    requires std::is_floating_point_v<T>
+  ALPAKA_FN_HOST_ACC ALPAKA_FN_INLINE T reduceRange(TAcc const& acc, T x) {
```
Comment (fwyzard) [inline HeterogeneousCore/AlpakaMath/interface/deltaPhi.h:11]: `reduceRange` is not a good name, because it doesn't convey any relation to the [-π, π].

File: DataFormats/Math/interface/alpaka/deltaPhi.h (hunk 2 above)
Comment (fwyzard) [reply DataFormats/Math/interface/alpaka/deltaPhi.h]: Note that `ALPAKA_FN_INLINE` is much stronger than `inline`, as it tries to force the compiler to actually inline the function.

The regular `inline` should be fine most of the time.

File: DataFormats/Math/interface/alpaka/deltaPhi.h (hunk 3 above)
Comment (fwyzard) [reply DataFormats/Math/interface/alpaka/deltaPhi.h]: Note that doing the computation in double precision may be quite expensive on most GPUs. For example it is 64 times more expensive than single precision on an NVIDIA L4.

File: HeterogeneousCore/AlpakaMath/interface/deltaPhi.h (hunk 4 above)
Comment (fwyzard) [reply HeterogeneousCore/AlpakaMath/interface/deltaPhi.h:11]: maybe `reducePhiRange` would be better ?

Comment (VourMa) [reply HeterogeneousCore/AlpakaMath/interface/deltaPhi.h:11]: I agree, and this was brought up by Matti as well. However, in this PR, I tried to stick to the current CMSSW convention which uses `reduceRange` in multiple places for keeping φ within [-π, π]:
https://github.com/search?q=repo%3Acms-sw%2Fcmssw%20reduceRange&type=code
Do we want to change just for this version of this function?

Maybe a global renaming to something clearer (`reducePhiRange` sounds good to me) would make sense. If at least some of us think it's useful and it doesn't turn out to be a rabbithole, I could try to propose it in a different PR for wider review. But I wouldn't connect it to this PR.

File: DataFormats/Math/interface/alpaka/deltaPhi.h (hunk 2 above)
Comment (VourMa) [reply DataFormats/Math/interface/alpaka/deltaPhi.h]: > The regular inline should be fine most of the time.

Should I be switching to that in this case? I am not sure whether you were suggesting that I apply this change or it was a more general comment.

File: DataFormats/Math/interface/alpaka/deltaPhi.h (hunk 3 above)
Comment (VourMa) [reply DataFormats/Math/interface/alpaka/deltaPhi.h]: I will go ahead and apply the suggestion by Matti to fix this:
```constexpr T o2pi = T{1.} / (T{2.} * std::numbers::pi_v<T>);```

**Comment by slava77** [comment]
@fwyzard 
should this be formally assigned to `heterogenous`? the PR currently has only `orp` atm

**Comment by VourMa** [comment]
I updated the PR applying @fwyzard's suggestions. I didn't act on renaming the function, as the preferred way to proceed is not clear to me. If everything else looks in order, I would have this PR merged at this stage, and I will come back with a follow-up PR (another one is in line anyways to actually use these functions in LST) in case we decide on renaming, either just this definition or all of them.

**Comment by slava77** [comment]
@cmsbuild please test

... btw, are the GPU unit tests executed this way, or does it require `enable gpu`?

**Comment by fwyzard** [comment]
enable gpu

is as far as I know required to run the tests on NVIDIA (and AMD ?) gpus

**Comment by makortel** [comment]
> should this be formally assigned to `heterogenous`? the PR currently has only `orp` atm

The new package anyway needs to be assigned an l2 category (with a PR to https://github.com/cms-sw/cms-bot/), after which the bot would do the assignment. The package being in `HeterogeneousCore` would imply at least `heterogeneous`. Given that `DataFormats/Math` is reviewed by `reconstruction`, @cms-sw/reconstruction-l2 do you want to review the `HeterogeneousCore/AlpakaMath` as well?

(I'm assuming @fwyzard isn't objecting the `HeterogeneousCore/AlpakaMath` package)

**Comment by fwyzard** [comment]
> (I'm assuming @fwyzard isn't objecting the `HeterogeneousCore/AlpakaMath` package)

no objections

**Comment by makortel** [comment]
> > (I'm assuming @fwyzard isn't objecting the `HeterogeneousCore/AlpakaMath` package)
> 
> no objections

I opened https://github.com/cms-sw/cms-bot/pull/2413 to add the package to `heterogeneous`.

File: HeterogeneousCore/AlpakaMath/interface/deltaPhi.h [hunk 5]
```diff
@@ -0,0 +1,36 @@
+#ifndef HeterogeneousCore_AlpakaMath_deltaPhi_h
+#define HeterogeneousCore_AlpakaMath_deltaPhi_h
+
+#include <alpaka/alpaka.hpp>
+
+namespace cms::alpakatools {
+
+  // reduce to [-pi,pi]
+  template <typename TAcc, typename T>
+    requires std::is_floating_point_v<T>
```
Comment (makortel) [inline HeterogeneousCore/AlpakaMath/interface/deltaPhi.h:10]: One alternative that I forgot to mention earlier would be
```suggestion
  template <typename TAcc, std::floating_point T>
```
(just mentioning, not asking to change unless you'd want to)

File: HeterogeneousCore/AlpakaMath/test/alpaka/testDeltaPhi.dev.cc [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+};
+
+template <typename T>
+void testPhiFuncs(uint32_t size, std::vector<double> const& res) {
+  // get the list of devices on the current platform
+  auto const& devices = cms::alpakatools::devices<Platform>();
+  if (devices.empty()) {
+    std::cerr << "No devices available for the " EDM_STRINGIZE(ALPAKA_ACCELERATOR_NAMESPACE) " backend, "
+      "the test will be skipped.\n";
+    exit(EXIT_FAILURE);
```
Comment (makortel) [inline HeterogeneousCore/AlpakaMath/test/alpaka/testDeltaPhi.dev.cc:51]: Since you are using catch2, this would be better expressed as
```suggestion
    FAIL("No devices available for the " EDM_STRINGIZE(ALPAKA_ACCELERATOR_NAMESPACE) " backend, "
      "the test will be skipped.");
```

File: HeterogeneousCore/AlpakaMath/test/alpaka/testDeltaPhi.dev.cc [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+    std::cout << "...on " << alpaka::getName(device) << "\n";
+    Queue queue(device);
+
+    auto c_h = make_host_buffer<T[]>(queue, size);
+    alpaka::memset(queue, c_h, 0.);
+    auto c_d = make_device_buffer<T[]>(queue, size);
+    alpaka::memset(queue, c_d, 0.);
+
+    alpaka::enqueue(queue,
+                    alpaka::createTaskKernel<Acc1D>(WorkDiv1D{1u, 1u, 1u}, phiFuncsUnitTestsKernel(), c_d.data()));
```
Comment (makortel) [inline HeterogeneousCore/AlpakaMath/test/alpaka/testDeltaPhi.dev.cc:64]: How about
```suggestion
    alpaka::exec<Acc1D>(queue, WorkDiv1D{1u, 1u, 1u}, phiFuncsUnitTestsKernel(), c_d.data());
```
?

File: HeterogeneousCore/AlpakaMath/test/alpaka/testDeltaPhi.dev.cc [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+    alpaka::memcpy(queue, c_h, c_d);
+    alpaka::wait(queue);
+
+    constexpr T eps = 1.e-5;
+    for (size_t i = 0; i < size; ++i) {
+      ok(c_h.data()[i], static_cast<T>(res[i]), eps);
```
Comment (makortel) [inline HeterogeneousCore/AlpakaMath/test/alpaka/testDeltaPhi.dev.cc:70]: In catch2 this check could be expressed as
```suggestion
      CHECK_THAT(c_h.data()[i], Catch::Matchers::WithinAbs(res[i], eps));
```

**Review by makortel — COMMENTED** [review]
Thanks for the unit test, here a few more improvement suggestions.

File: HeterogeneousCore/AlpakaMath/interface/deltaPhi.h [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+  ALPAKA_FN_HOST_ACC inline T reduceRange(TAcc const& acc, T x) {
+    constexpr T o2pi = T{1.} / (T{2.} * std::numbers::pi_v<T>);
```
Comment (makortel) [inline HeterogeneousCore/AlpakaMath/interface/deltaPhi.h:12]: It would be good to add
```cpp
#include <numbers>
```

**Comment by VourMa** [comment]
Thanks for the extra suggestions, @makortel, I applied all of them.

**Comment by slava77** [comment]
@cmsbuild please test

File: HeterogeneousCore/AlpakaMath/interface/deltaPhi.h [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+
+namespace cms::alpakatools {
+
+  // reduce to [-pi,pi]
+  template <typename TAcc, std::floating_point T>
+  ALPAKA_FN_HOST_ACC inline T reduceRange(TAcc const& acc, T x) {
+    constexpr T o2pi = T{1.} / (T{2.} * std::numbers::pi_v<T>);
+    if (alpaka::math::abs(acc, x) <= std::numbers::pi_v<T>)
+      return x;
+    T n = alpaka::math::round(acc, x * o2pi);
```
Comment (fwyzard) [inline HeterogeneousCore/AlpakaMath/interface/deltaPhi.h:14]: Just curious: why not simply
```suggestion
    T n = alpaka::math::round(acc, x / (T{2} * std::numbers::pi_v<T>));
```
or
```suggestion
    T n = alpaka::math::round(acc, x * std::numbers::inv_pi_v<T> * T{0.5});
```
?

Comment (VourMa) [reply HeterogeneousCore/AlpakaMath/interface/deltaPhi.h:14]: As initially mentioned, this was meant to be an alpaka-adjusted copy of the `reco::reduceRange` function
https://github.com/cms-sw/cmssw/blob/a475844e9e68627f9a367b74a2beb482f965d3a7/DataFormats/Math/interface/deltaPhi.h#L18-L24
hence I didn't apply any other simplifications, apart from the ones suggested in this PR.

Comment (fwyzard) [reply HeterogeneousCore/AlpakaMath/interface/deltaPhi.h:14]: ok

**Comment by VourMa** [comment]
Anything still pending for this PR? Or can it be merged so that I can continue with the follow up PRs?

**Comment by makortel** [comment]
> Anything still pending for this PR? Or can it be merged so that I can continue with the follow up PRs?

We can't formally sign until https://github.com/cms-sw/cms-bot/pull/2413 gets merged and the package assigned

**Comment by smuzaffar** [comment]
https://github.com/cms-sw/cms-bot/pull/2413 is merged now

**Comment by fwyzard** [comment]
+heterogeneous

**Comment by mandrenguyen** [comment]
+1
